cilium/CFP-30788-managing-hubble-metrics-cardinality.md

```diff
@@ -0,0 +1,130 @@
+
+- When the cluster is using a managed prometheus offering where the prometheus configuration is managed by the cloud provider and custom ingestion rules cannot be applied.
+- When a high scale cluster is generating high cardinality of metrics putting pressure on the prometheus ingestion, storage and query performance.
+- When the cluster admin wants to dynamically control the cardinality of the metrics based on the cluster state without having to restart cilium agent or prometheus server. 
+
+Such issues present a unique challenge to cluster admins and developers. By creating a mechanism to manage the cardinality of Hubble metrics dynamically based on the cluster state would help in addressing the above challenges. Recent change for [Hubble Flows dynamic exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) has inspired this proposal.
+
+## Goals
+
+* Ability to manage the cardinality of Hubble metrics dynamically.
+* Should be able to control labels and dimensions of the metrics.
+
+
+## Non-Goals
+
+* Recommendations for Prometheus configuration at ingestion time.
+
+## Proposal
+
+### Overview
+
+We propose a new config map wrapped functionality in Hubble metrics which can be used to manage the cardinality, dimensions and targets of the metrics. Drawing from [Hubble Flows dynamic exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) which uses a standalone config map to manage exporting of flows, we propose a similar mechanism. 
+
+Existing behavior of reading cilium-config for `hubble-metrics` be preserved for backward compatibility. 
+
+A new config called `enable-dynamic-hubble-metrics` be introduced which when set to true, will enable the dynamic management of Hubble metrics. When this config is set to true, the cilium agent will read the `hubble-metrics-config` config map to get the configuration for the metrics. When this config is set to false, the cilium agent will read the `cilium-config` for `hubble-metrics` configuration.
+
+### Config Map
+
+A new config map called `hubble-metrics-config` will be introduced which will have the following fields:
+
+- `enabled`: Boolean value to enable/disable the dynamic management of Hubble metrics.
+- `metrics`: List of metrics to be managed. Each metric will have the following fields:
+  - `name`: Name of the metric.
+  - `contextoptions`: List of [context options](https://docs.cilium.io/en/latest/observability/metrics/#context-options) to be applied to the metric.
+  - `labels`: List of labels to be applied to the metric.
+  - `includeFilters` : List of include flow filters to be applied to this particular metric.
+  - `excludeFilters` : List of exclude flow filters to be applied to this particular metric.
+
+### Example
+    
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+    name: hubble-metrics-config
+    namespace: kube-system
+data:
+    hubblemetrics.yaml: |
+        hubbleMetrics:
+            enabled: true
```

Comment:
Do we need the `enabled` field? There's already controls to disable Hubble entirely. So, even if this CM is applied and Hubble is disabled, the reconcile is basically no-op.

Reply:
I think it's useful to have an enabled/disabled option here, since the only other way to disable is to not configure metrics in the first place. Having a config option to enable/disable means something else managing the configuration out-of-band (eg: a separate add-on control plane for integrating with a hosted K8s solution), means that it could be toggled on/off without restarting cilium. (in theory)

Reply:
Can you expand on this a bit, i am thinking this enabled/disabled could be a redundant field, source of truth for if this feature is enabled will be in `cilium-config` right ? Even if admin disables here in this configmap, hubble willl read the `enable-dynamic-hubble-metrics` in `cilium-config` and keep the feature enabled?

I am inclining towards removing this flag from new configmap.

Reply:
Basically, imagine your a K8s as a service provider and you have cilium as part of your offering. Users cannot control the cilium install via helm, or by modify the cilium-config, etc. 

So as part of your hosted k8s offering, you have a separate microservice which creates the `hubblemetrics.yaml` configmap based on the settings the user has set in your web UI. One option is to disable metrics which would set `hubbleMetrics.disabled=true` in the configmap and cilium will not configure any Hubble Metrics. Later on, the user can enable Hubble Metrics, which will change the configmap to have `hubbleMetrics.disabled=false`. 

This means the Hubble Metrics feature can be completely controlled without having to restart cilium or anything.

Reply:
Is this enable/disable of metrics overall, or _dynamic_ metrics?

Reply:
below is my assumption on how the ordering of config maps will work:

1.  `cilium-config` : `enable-hubble `- will enable Hubble itself
2. `cilium-config` :  `enable-dynamic-hubble-metrics`  - will make any metrics related config in cilium-config obsolete and will consider `hubble-metrics-config` as the only source of truth
3. `hubble-metrics-config`: `enabled/disabled` will enable/disable hubble metrics 

Correct me if my thinking is incorrect.  

Reply:
That matches my expectations